[Solomon Ng] iP

solomonng2001 opened the pull request:
# Banter

> “You may delay, but time will not.” – Benjamin Franklin ([Source](https://www.socratic-method.com/quote-meanings/benjamin-franklin-you-may-delay-but-time-will-not#:~:text=%22You%20may%20delay%2C%20but%20time%20will%20not.%22%20These,off%20tasks%2C%20time%20continues%20to%20pass%20by%20relentlessly.))

Banter helps you organise your time and keep track of the things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ LIGHTNING-FAST: Get organised in seconds!

All you need to do is,
1. download it from here
2. double-click it
3. add your tasks
4. Sit back and let it manage your tasks for you :grin:

And it is **FREE!**

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events
- [ ] Reminders (coming soon)

If you are familiar with Java, you might be interested with the source code. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Banter banter = new Banter();
        banter.start();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

solomonng2001 committed:
rename-chatbot-and-add-greet-and-exit-responses

solomonng2001 committed:
add-echo-logic-and-restructure-directories

solomonng2001 committed:
(1). Addition of task list
(2). Use enums for commands input by user
(3). Refactor MessageBox to handle message strings instead of arrays

Regarding (2), there are only specific set of possible commands being
input by user.

Regarding (3), simpler to use strings as message body, and indent each
line by using a parser.

solomonng2001 committed:
add-mark-and-unmark-tasks

solomonng2001 committed:
Addition of Todos, Events, Deadlines

(1) Inheritance of Todos, Events, Deadlines from Task
(2) Parsing of new commands
(3) Addition of new commands to enum and responses

solomonng2001 committed:
add-error-handling

solomonng2001 committed:
add-delete-task

solomonng2001 committed:
redesign-parser-to-have-controller-functions

solomonng2001 committed:
add-input-output-redirection-testing

solomonng2001 committed:
add-input-output-testing

solomonng2001 committed:
add-storage-and-save-tasklist-data

solomonng2001 committed:
Merge branch 'branch-Level-7'

solomonng2001 committed:
add-dates-and-times-formatting-storage-parsing

solomonng2001 committed:
Merge branch 'branch-Level-8'

solomonng2001 committed:
extract-ui-messages-card

solomonng2001 committed:
envelope-all-classes-in-banter-package

solomonng2001 committed:
Merge branch 'add-gradle-support'

solomonng2001 committed:
add-gradle-support

solomonng2001 committed:
add-tests-cases-for-todo-class

solomonng2001 committed:
add-jar-file-via-gradle

solomonng2001 committed:
add-javadoc-comments-to-most-public-methods

solomonng2001 committed:
add-test-cases-for-event-class

solomonng2001 committed:
add-find-keyword-command-and-function

solomonng2001 committed:
Add JavaDoc comments

This commit implements the A-JavaDoc task, adding JavaDoc comments
to enhance code documentation. Header comments have been added to at
least half of the non-private classes and methods as required by the
minimal task.

This improvement enhances code readability and provides better
documentation for future maintenance.

solomonng2001 committed:
Follow the Coding Standard

This commit addresses the A-CodingStandard task, where the code has
been adjusted to comply with the specified coding standard. Moving
forward, any new code introduced will adhere to this standard,
ensuring a consistent and maintainable codebase.

This is to improve code readability and maintainability.

solomonng2001 committed:
Implement Task Search Functionality

This commit introduces the find command, allowing users to search
for tasks based on a specified keyword. The example provided
demonstrates how the command can be used to find tasks containing the
keyword book. The matching tasks are displayed in a formatted manner,
enhancing the user's ability to locate relevant information within the
task list.

This will increase the functionality of Task List.

solomonng2001 committed:
Add JavaDoc comments

This commit implements the "A-JavaDoc" task, adding JavaDoc comments
to enhance code documentation. Header comments have been added to at
least half of the non-private classes and methods as required by the
minimal task.

This improvement enhances code readability and provides
better documentation for future maintenance.

Merge branch 'refs/heads/A-JavaDoc'

solomonng2001 committed:
Follow the Coding Standard

This commit addresses the "A-CodingStandard" task, where the code has
been adjusted to comply with the specified coding standard. Moving
forward, any new code introduced will adhere to this standard,
ensuring a consistent and maintainable codebase.

This ensures readability and maintainability.

Merge branch 'refs/heads/A-CodingStandard'

solomonng2001 committed:
Fix merge conflicts when merging Level-9 to Master

This commit introduces the find command, allowing users to search
for tasks based on a specified keyword. The example provided
demonstrates how the command can be used to find tasks containing
the keyword book. The matching tasks are displayed in a formatted
manner, enhancing the user's ability to locate relevant information
within the task list.

This increase the functionality of Task List, allowing searching.

solomonng2001 committed:
Fix invalid storage format bug

This commit fixes the incorrect storage formatting, that led to an
IndexOutOfBoundsError.

dwangwk commented on src/main/java/seedu/banter/Parser.java:
```diff
@@ -0,0 +1,267 @@
+
+
+/**
+ * Represents a parser that parses user input and functions as the app controller.
+ */
+public class Parser {
```
Great Parser I like it 👍 .

dwangwk commented on src/main/java/seedu/banter/errors/Errors.java:
```diff
@@ -0,0 +1,29 @@
+
+import seedu.banter.ui.Ui;
+
+public class Errors {
```
While the code is relatively self-explanatory, adding comments on top of the Errors class or below before each constant declaration to describe the purpose of the class and each constant could enhance readability? Other than that, this Errors class is great!

dwangwk commented on src/main/java/seedu/banter/tasks/TaskList.java:
```diff
@@ -0,0 +1,179 @@
+import seedu.banter.errors.InvalidTaskNumberUsageError;
+import seedu.banter.ui.Ui;
+
+/**
```
Perhaps you can consider adding more detailed documentation, such as including information about its fields, methods, and their purposes!

dwangwk reviewed (commented):
Overall looks good to me! Your project goes above and beyond the project specifications. Handling of Errors and Exceptions are really well done.

solomonng2001 committed:
Modify style to follow coding standard

This commit modifies the style to follow to java coding standard, as
outlined by the course's checkstyle. This is for readability and
maintainability.

solomonng2001 committed:
Add GUI user interface

Addition of GUI user interphase will make the application more usable.

solomonng2001 committed:
Merge branch 'branch-Level-10'

Addition of GUI user interface will make application mroe usable.

solomonng2001 committed:
Enable find task to accept multiple keywords

Enabling the find task function to accept multiple keywords, using
variable arguments, will broaden the functionality of our search.

theman-oj10 commented on src/main/java/seedu/banter/Parser.java:
```diff
@@ -0,0 +1,267 @@
+            String input = scanner.nextLine();
+            try {
+                command = getCommandType(input);
+                switch (command) {
```
Minor nitpick but the cases should have the same indentation as the switch statements.

theman-oj10 commented on src/main/java/seedu/banter/Parser.java:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    private void parseTodo(String input) throws InvalidBanterUsageError {
+        String[] tokens = getTokens(input);
```
Just clarifying, what do tokens mean? I might be wrong but perhaps the naming isn't too straightforward?

theman-oj10 commented on src/main/java/seedu/banter/Parser.java:
```diff
@@ -0,0 +1,267 @@
+        if (startStr.isEmpty()) {
+            throw Errors.MISSING_EVENT_START_ERROR;
+        }
+        LocalDateTime start = DateTime.getDateTimeFromUserInput(startStr);
```
The naming for methods is very good in terms of them being self-explanatory

theman-oj10 commented on src/main/java/seedu/banter/Storage.java:
```diff
@@ -0,0 +1,120 @@
+            }
+            return taskList;
+        } catch (FileNotFoundException e) {
+            handleFileNotFound();
```
Good job on the naming of error handling methods and overall good job on abstracting out all your error handling.

theman-oj10 commented on src/main/java/seedu/banter/Storage.java:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+    
+    private String toStorageFormat(LocalDateTime dateTime) {
```
Noticing multiple methods with the same name but handling different parameters, good use of Overloading! This is good since these are helper methods.

theman-oj10 commented on src/main/java/seedu/banter/tasks/Event.java:
```diff
@@ -0,0 +1,75 @@
+     * Returns the start time of an Event task.
+     * @return Start time of an Event task.
+     */
+    public LocalDateTime getStart() {
```
Perhaps name this getStartTime()? to make it a little clearer. Can do the same for getEndTime().

theman-oj10 reviewed (commented):
LGTM, minor nits with some of the naming. Overall very well done!

solomonng2001 committed:
Add assertions: control flow and class invariants

Lack of assertions.

Assertions will help us document important invariants at various
points of our code.

Let's,
* add class invariants to check the values assigned to the class
variables through the constructor parameters
* add control flow invariants to check correctness of our control
flow and ensure that there are no unintended paths taken

solomonng2001 committed:
Adopt self-explanatory names

Some method names are not explicit enough.

Having self-explanatory names will help us make readability for
future developers easier.

Let's,
* Ensure that all method, class and variable names are
self-explanatory.

solomonng2001 committed:
Merge pull request #1 from solomonng2001/branch-A-Assertions

Add assertions: control flow and class invariants

solomonng2001 committed:
Merge pull request #2 from solomonng2001/branch-A-CodeQuality

Adopt self-explanatory names

solomonng2001 committed:
Use streams where beneficial

Lack of streams where usage might be beneficial.

In some cases, use of streams can make code more concise and readable.

Let's,
* Replace code with streams, in cases where streams will improve
the readability and conciseness of the code.

solomonng2001 committed:
Add sorting of task list

Tasks are ordered by when user adds it to the list.

Sorting by date and time will allow user to work on more urgent tasks
first.

Let's,
* Sort tasks by date and time whenever a task list is stored
or printed

solomonng2001 committed:
Improve the GUI

Dialog bubble did not exist and text was floating around.

Improving the GUI will make the text more readable for the user.

Let's
* Design a text bubble to hold the text.
* Touch up on the display of images.

solomonng2001 committed:
Add of test cases for deadline

No test cases for deadline.

More test cases will help reduce regressions in the future.

Let's
* Add test cases for deadline.
* Ensure no checkstyle violations.

solomonng2001 committed:
Add more error handling

Error handling is limited and existence of println statements.

Addition of error handling will reduce chances of bugs and removing
println will allow errors to be displayed by the GUI.

Let's
* Add load task list error handling.
* Add save task list error handling.
* Add date time error handling.
* Remove error handling by println and return string to GUI instead.

solomonng2001 committed:
Improve GUI

Banter's mascot was not aligned right.

Aligning Banter's mascot right will make the GUI more crisp.

Let's
* Align our GUI items properly.

solomonng2001 committed:
Remove unused methods

Unused methods are present.

Removing unused methods will make project more maintainable and concise.

Let's
* Remove unused methods.

solomonng2001 committed:
Make find command more dynamic

Find command can only search by a single keyword.

Making find command more dynamic will allow better search function.

Let's
* Enable find command to search by multiple keywords.

solomonng2001 committed:
Add user guide

User guide does not exist.

A user guide will provide convenient usage information to users and
warn them of improper usage.

Let's
* Add a detailed user guide with command summary, features, introduction
and quick start.

solomonng2001 committed:
Edit user guide

Some images in user guide are not positioned properly.

Poorly positioned images will make the user guide not look crisp.

Let's
* Properly reposition items in the user guide.

solomonng2001 committed:
Add link from user guide to release

Link from user guide to release does not exist.

Hyperlinks will make it easy for new users to download the app.

Let's
* Add hyperlinks that will convenience new users.

solomonng2001 committed:
Correct markdown list

Markdown list was wrongly sequenced.

Ensuring that list is sequenced properly will ensure user will not be
confused when reading the guide.

Let's
* ensure all lists are properly sequenced.

solomonng2001 committed:
Update README.md files

Root README.md file was not updated.

Updating README.md is necessary to ensure that readers of the
repository know how to set up the project.

Let's
* Update all README.md to ensure that they have the latest information.

solomonng2001 committed:
Implement continuous integration

Continuous integration does not exist.

Having continuous integration will improve code quality by detecting
issues earlier and provide a convenient and automated way to build,
test and validate code.

Let's
* Set up continuous integration processes.

solomonng2001 committed:
Add css for dialogbox

Dialogbox does not have css design.

Adding css design will make the overall ui more crisp.

Let's
* Add css to ui elements.

solomonng2001 committed:
Remove markdown errors in user guide

Markdown errors exist in user guide.

Removing accidental presence of html or other markdown appearing in
user guide will ensure a crisp formatting.

Let's,
* Remove html appearing in user guide github pages.
